src/main/java/org/spongepowered/common/world/DimensionManager.java

```diff
@@ -24,336 +24,982 @@
+                    Sponge.getServer()).getDataFixer()).saveWorldInfo((WorldInfo) properties);
         }
+        // No return values or exceptions so can only assume true.
+        return true;
+    }
+
+    public static void unloadQueuedWorlds() {
+        while (unloadQueue.peek() != null) {
+            final WorldServer worldServer = unloadQueue.poll();
+
+            unloadWorld(worldServer);
+        }
+
+        unloadQueue.clear();
+    }
+
+    // TODO Result
+    public static boolean unloadWorld(WorldServer worldServer) {
+        checkNotNull(worldServer);
+        final MinecraftServer server = (MinecraftServer) Sponge.getServer();
 
-        ArrayList<WorldServer> tmp = new ArrayList<>();
-        if (worlds.get(0) != null) {
-            tmp.add(worlds.get(0));
+        if (worldByDimensionId.containsValue(worldServer)) {
+            if (!worldServer.playerEntities.isEmpty()) {
+                return false;
+            }
+
+            try {
+                saveWorld(worldServer);
+            } catch (MinecraftException e) {
+                e.printStackTrace();
+            }
+            finally {
+                final int dimensionId = ((IMixinWorld) worldServer).getDimensionId();
+                worldByDimensionId.remove(dimensionId);
+                ((IMixinMinecraftServer) server).getWorldTickTimes().remove(dimensionId);
+                SpongeImpl.getLogger().info("Unloading dimension {} ({})", dimensionId, worldServer.getWorldInfo().getWorldName());
+            }
+
+            SpongeImpl.postEvent(SpongeEventFactory.createUnloadWorldEvent(Cause.of(NamedCause.source(server)), (org.spongepowered.api.world.World)
+                    worldServer));
+
+            return true;
         }
-        if (worlds.get(-1) != null) {
-            tmp.add(worlds.get(-1));
+        return false;
+    }
+
+    private static void saveWorld(WorldServer worldServer) throws MinecraftException {
+        final MinecraftServer server = (MinecraftServer) Sponge.getServer();
+        final org.spongepowered.api.world.World apiWorld = (org.spongepowered.api.world.World) worldServer;
+
+        Sponge.getEventManager().post(SpongeEventFactory.createSaveWorldEventPre(Cause.of(NamedCause.source(server)), apiWorld));
+
+        worldServer.saveAllChunks(true, null);
+        worldServer.flush();
+
+        Sponge.getEventManager().post(SpongeEventFactory.createSaveWorldEventPost(Cause.of(NamedCause.source(server)), apiWorld));
+    }
+
+    public static Collection<WorldProperties> getUnloadedWorlds() throws IOException {
+        final Optional<Path> optCurrentSavesDir = getCurrentSavesDirectory();
+        checkState(optCurrentSavesDir.isPresent(), "Attempt made to get unloaded worlds too early!");
+        final DirectoryStream<Path> stream = Files.newDirectoryStream(optCurrentSavesDir.get(), LEVEL_AND_SPONGE);
```

Comment:
Make sure to close this stream, e.g. by simply putting it into a try-with-resources statement (without a `catch` clause)